File: rng_tools/src/pkm/nature.rs
use num_enum::FromPrimitive;
use serde::{Deserialize, Serialize};
use tsify_next::Tsify;

#[derive(Clone, Copy, Debug, Eq, PartialEq, FromPrimitive, Tsify, Serialize, Deserialize)]
#[tsify(into_wasm_abi, from_wasm_abi)]
#[repr(u8)]
pub enum Nature {
    #[num_enum(default)]
    Hardy = 0,
    Lonely = 1,
    Brave = 2,
    Adamant = 3,
    Naughty = 4,
    Bold = 5,
    Docile = 6,
    Relaxed = 7,
    Impish = 8,
    Lax = 9,
    Timid = 10,
    Hasty = 11,
    Serious = 12,
    Jolly = 13,
    Naive = 14,
    Modest = 15,
    Mild = 16,
    Quiet = 17,
    Bashful = 18,
    Rash = 19,
    Calm = 20,
    Gentle = 21,
    Sassy = 22,
    Careful = 23,
    Quirky = 24,
}

impl Nature {
    pub fn from_pid(pid: u32) -> Self {
        ((pid % 25) as u8).into()
    }
}

impl Default for Nature {
    fn default() -> Self {
        Self::Hardy
    }
}

#[derive(Copy, Clone)]
pub enum NatureFactor {
  More,
  Less,
  Equal,
}

pub struct NatureStatFactor {
    pub atk: NatureFactor,
    pub def: NatureFactor,
    pub spa: NatureFactor,
    pub spd: NatureFactor,
    pub spe: NatureFactor,
}

pub const NATURE_STAT_FACTORS:[NatureStatFactor;25] = [
    /*Hardy */    NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Equal, spe: NatureFactor::Equal, spa: NatureFactor::Equal, spd: NatureFactor::Equal },
    /*Lonely */   NatureStatFactor { atk: NatureFactor::More, def: NatureFactor::Less, spe: NatureFactor::Equal, spa: NatureFactor::Equal, spd: NatureFactor::Equal },
    /*Brave */    NatureStatFactor { atk: NatureFactor::More, def: NatureFactor::Equal, spe: NatureFactor::Less, spa: NatureFactor::Equal, spd: NatureFactor::Equal },
    /*Adamant */  NatureStatFactor { atk: NatureFactor::More, def: NatureFactor::Equal, spe: NatureFactor::Equal, spa: NatureFactor::Less, spd: NatureFactor::Equal },
    /*Naughty */  NatureStatFactor { atk: NatureFactor::More, def: NatureFactor::Equal, spe: NatureFactor::Equal, spa: NatureFactor::Equal, spd: NatureFactor::More },
    /*Bold */     NatureStatFactor { atk: NatureFactor::Less, def: NatureFactor::More, spe: NatureFactor::Equal, spa: NatureFactor::Equal, spd: NatureFactor::Equal },
    /*Docile */   NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Equal, spe: NatureFactor::Equal, spa: NatureFactor::Equal, spd: NatureFactor::Equal },
    /*Relaxed */  NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::More, spe: NatureFactor::Less, spa: NatureFactor::Equal, spd: NatureFactor::Equal },
    /*Impish */   NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::More, spe: NatureFactor::Equal, spa: NatureFactor::Less, spd: NatureFactor::Equal },
    /*Lax */      NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::More, spe: NatureFactor::Equal, spa: NatureFactor::Equal, spd: NatureFactor::More },
    /*Timid */    NatureStatFactor { atk: NatureFactor::Less, def: NatureFactor::Equal, spe: NatureFactor::More, spa: NatureFactor::Equal, spd: NatureFactor::Equal },
    /*Hasty */    NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Less, spe: NatureFactor::More, spa: NatureFactor::Equal, spd: NatureFactor::Equal },
    /*Serious */  NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Equal, spe: NatureFactor::Equal, spa: NatureFactor::Equal, spd: NatureFactor::Equal },
    /*Jolly */    NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Equal, spe: NatureFactor::More, spa: NatureFactor::Less, spd: NatureFactor::Equal },
    /*Naive */    NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Equal, spe: NatureFactor::More, spa: NatureFactor::Equal, spd: NatureFactor::Less },
    /*Modest */   NatureStatFactor { atk: NatureFactor::Less, def: NatureFactor::Equal, spe: NatureFactor::Equal, spa: NatureFactor::More, spd: NatureFactor::Equal },
    /*Mild */     NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Less, spe: NatureFactor::Equal, spa: NatureFactor::More, spd: NatureFactor::Equal },
    /*Quiet */    NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Equal, spe: NatureFactor::Less, spa: NatureFactor::More, spd: NatureFactor::Equal },
    /*Bashful */  NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Equal, spe: NatureFactor::Equal, spa: NatureFactor::Equal, spd: NatureFactor::Equal },
    /*Rash */     NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Equal, spe: NatureFactor::Equal, spa: NatureFactor::More, spd: NatureFactor::Less },
    /*Calm */     NatureStatFactor { atk: NatureFactor::Less, def: NatureFactor::Equal, spe: NatureFactor::Equal, spa: NatureFactor::Equal, spd: NatureFactor::More },
    /*Gentle */   NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Less, spe: NatureFactor::Equal, spa: NatureFactor::Equal, spd: NatureFactor::More },
    /*Sassy */    NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Equal, spe: NatureFactor::Less, spa: NatureFactor::Equal, spd: NatureFactor::More },
    /*Careful */  NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Equal, spe: NatureFactor::Equal, spa: NatureFactor::Less, spd: NatureFactor::More },
    /*Quirky */   NatureStatFactor { atk: NatureFactor::Equal, def: NatureFactor::Equal, spe: NatureFactor::Equal, spa: NatureFactor::Equal, spd: NatureFactor::Equal },
];

File: src/types/nature.ts
import { Nature } from "~/rngTools";
import type { Stat } from "./stat";
import * as tst from "ts-toolbelt";

export type NatureStat = tst.U.Exclude<Stat, "hp">;

export const natures = [
  "Hardy",
  "Lonely",
  "Brave",
  "Adamant",
  "Naughty",
  "Bold",
  "Docile",
  "Relaxed",
  "Impish",
  "Lax",
  "Timid",
  "Hasty",
  "Serious",
  "Jolly",
  "Naive",
  "Modest",
  "Mild",
  "Quiet",
  "Bashful",
  "Rash",
  "Calm",
  "Gentle",
  "Sassy",
  "Careful",
  "Quirky",
] as const satisfies Nature[];

export const getNaturesFromStatMoreLess = (
  statMore: NatureStat | "",
  statLess: NatureStat | "",
): Nature[] => {
  if (statMore === "" && statLess === "")
    return ["Hardy", "Docile", "Bashful", "Quirky", "Serious"];

  if (statMore === "atk") {
    if (statLess === "def") return ["Lonely"];
    if (statLess === "spa") return ["Adamant"];
    if (statLess === "spd") return ["Naughty"];
    if (statLess === "spe") return ["Brave"];
  }
  if (statMore === "def") {
    if (statLess === "atk") return ["Bold"];
    if (statLess === "spa") return ["Impish"];
    if (statLess === "spd") return ["Lax"];
    if (statLess === "spe") return ["Relaxed"];
  }
  if (statMore === "spa") {
    if (statLess === "atk") return ["Modest"];
    if (statLess === "def") return ["Mild"];
    if (statLess === "spd") return ["Rash"];
    if (statLess === "spe") return ["Quiet"];
  }
  if (statMore === "spd") {
    if (statLess === "atk") return ["Calm"];
    if (statLess === "def") return ["Gentle"];
    if (statLess === "spa") return ["Careful"];
    if (statLess === "spe") return ["Sassy"];
  }
  if (statMore === "spe") {
    if (statLess === "atk") return ["Timid"];
    if (statLess === "def") return ["Hasty"];
    if (statLess === "spa") return ["Jolly"];
    if (statLess === "spd") return ["Naive"];
  }
  return [];
};

//NO_PROD: improv syntax
export const getStatMoreLessFromNature = (
  nature: Nature,
): [NatureStat | "", NatureStat | ""] => {
  if (nature === "Hardy") return ["", ""];
  if (nature === "Lonely") return ["atk", "def"];
  if (nature === "Adamant") return ["atk", "spa"];
  if (nature === "Naughty") return ["atk", "spd"];
  if (nature === "Brave") return ["atk", "spe"];

  if (nature === "Bold") return ["def", "atk"];
  if (nature === "Docile") return ["", ""];
  if (nature === "Impish") return ["def", "spa"];
  if (nature === "Lax") return ["def", "spd"];
  if (nature === "Relaxed") return ["def", "spe"];

  if (nature === "Modest") return ["spa", "atk"];
  if (nature === "Mild") return ["spa", "def"];
  if (nature === "Bashful") return ["", ""];
  if (nature === "Rash") return ["spa", "spd"];
  if (nature === "Quiet") return ["spa", "spe"];

  if (nature === "Calm") return ["spd", "atk"];
  if (nature === "Gentle") return ["spd", "def"];
  if (nature === "Careful") return ["spd", "spa"];
  if (nature === "Quirky") return ["", ""];
  if (nature === "Sassy") return ["spd", "spe"];

  if (nature === "Timid") return ["spe", "atk"];
  if (nature === "Hasty") return ["spe", "def"];
  if (nature === "Jolly") return ["spe", "spa"];
  if (nature === "Naive") return ["spe", "spd"];
  if (nature === "Serious") return ["", ""];

  return ["", ""];
};
